src/diffusers/schedulers/scheduling_flow_match_euler_discrete.py

```diff
@@ -294,6 +295,10 @@ def step(
         sigma = self.sigmas[self.step_index]
         sigma_next = self.sigmas[self.step_index + 1]
 
+        if self.config.invert_sigmas:
+            print("inverting")
+            sigma, sigma_next = sigma_next, sigma
```

Comment:
Oh, nice catch! I did not notice that they were flipping `sigma_schedule` in the helper function.

I tried both your suggested changes with/without `invert_sigmas=True` (in the scheduler) and neither produce good videos.

From what I understand, the model considers timestep=0 as complete noise and timestep=1000 as clean latent (opposite of what is usually seen). Flipping the sigmas is not actually a null-op, because:
- when `invert_sigmas=False`, we do `(1 - actual_sigma_next) - (1 - actual_sigma)`, which is `actual_sigma - actual_sigma_next`
- when `invert_sigmas=True`, we do `(1 - actual_sigma) - (1 - actual_sigma_next)`, which is `actual_sigma_next - actual_sigma`

I don't think we will be able to do without the `invert_sigmas=True` change...

As I mentioned in our personal chat, I think genmo training code probably adds zero noise when timestep is 1000 and complete noise when timestep is 0. If this bug exists in their training code, which is the only possible explanation to me, the complete math works out as expected, except ofcourse the model learns to associate with noise levels inversely.

Reply:
oh I think it did not work because the timestep scale shift is applied before inverting the sigmas, the math still should holds though, i.e. if they use 0 - 1, we just have to set the timesteps that way, the model input should still be the timesteps (`sigmas` in our code) and the update math should still be the same 

I tried here https://github.com/huggingface/diffusers/commit/1069d210e1b9e84a366cdc7a13965626ea258178 ; the results are same but they both generate a same brownish videos so I'm not completely sure 😅 can you take a look?

Reply:
ok tested on the latest commit, I think it is the same

https://github.com/user-attachments/assets/6aaf4997-2c50-4037-95b2-d548fc02efee

Reply:
Okay yep this works, thanks! I think we will still need the `invert_sigmas` config and both of these implementations are same math-wise. But, still quite weird that timestep 1000 is clean latent and timestep 0 is noise 🤔

Reply:
Do we need `invert_sigmas` in the config? Can't we just delete this line as @yiyixuxu  suggested? 
https://github.com/huggingface/diffusers/pull/9769#discussion_r1817557791

Reply:
We will need invert sigmas one way or another (whether we do 1000 - timestep currently, or reverse sigmas in the scheduler as in her shared commit). I'll cherry pick it and add here

Reply:
Oh, what I meant is we don't flip them in `linear_quadtratic_schedule`. Wouldn't that be equivalent? Then we don't need to add an additional parameter to the scheduler config?

Reply:
So, normally one would pass timestep `1000` at the first denoising step, then something like `9xx` and then `8xx` and so on, right? Here you must pass timesteps in reverse order - so `0` first, `1xx` second... - because I believe the model has learnt to associate timestep condition `0` to noisy image and `1000` to clean latent. My comment [here](https://github.com/huggingface/diffusers/pull/9769#discussion_r1817601945) should possibly help with any confusion, but LMK if it's not clear.

Both mine `1000 - timestep` approach and YiYi's flipping of sigmas in the scheduler.set_timestep approach are equivalent. We just need to ensure that the order in which the model sees timesteps is 0 -> 1000 and not 1000 -> 0

Reply:
Are these not equivalent? 

```
import torch
import numpy as np
from diffusers import FlowMatchEulerDiscreteScheduler

# from: https://github.com/genmoai/models/blob/075b6e36db58f1242921deff83a1066887b9c9e1/src/mochi_preview/infer.py#L77
def _dont_flip_linear_quadratic_schedule(num_steps, threshold_noise, linear_steps=None):
    if linear_steps is None:
        linear_steps = num_steps // 2
    linear_sigma_schedule = [i * threshold_noise / linear_steps for i in range(linear_steps)]
    threshold_noise_step_diff = linear_steps - threshold_noise * num_steps
    quadratic_steps = num_steps - linear_steps
    quadratic_coef = threshold_noise_step_diff / (linear_steps * quadratic_steps**2)
    linear_coef = threshold_noise / linear_steps - 2 * threshold_noise_step_diff / (quadratic_steps**2)
    const = quadratic_coef * (linear_steps**2)
    quadratic_sigma_schedule = [
        quadratic_coef * (i**2) + linear_coef * i + const for i in range(linear_steps, num_steps)
    ]
    sigma_schedule = linear_sigma_schedule + quadratic_sigma_schedule + [1.0]
    # sigma_schedule = [1.0 - x for x in sigma_schedule]
    return np.array(sigma_schedule)

# from: https://github.com/genmoai/models/blob/075b6e36db58f1242921deff83a1066887b9c9e1/src/mochi_preview/infer.py#L77
def _flip_linear_quadratic_schedule(num_steps, threshold_noise, linear_steps=None):
    if linear_steps is None:
        linear_steps = num_steps // 2
    linear_sigma_schedule = [i * threshold_noise / linear_steps for i in range(linear_steps)]
    threshold_noise_step_diff = linear_steps - threshold_noise * num_steps
    quadratic_steps = num_steps - linear_steps
    quadratic_coef = threshold_noise_step_diff / (linear_steps * quadratic_steps**2)
    linear_coef = threshold_noise / linear_steps - 2 * threshold_noise_step_diff / (quadratic_steps**2)
    const = quadratic_coef * (linear_steps**2)
    quadratic_sigma_schedule = [
        quadratic_coef * (i**2) + linear_coef * i + const for i in range(linear_steps, num_steps)
    ]
    sigma_schedule = linear_sigma_schedule + quadratic_sigma_schedule + [1.0]
    sigma_schedule = [1.0 - x for x in sigma_schedule]
    return np.array(sigma_schedule)

scheduler = FlowMatchEulerDiscreteScheduler()
sigmas = _dont_flip_linear_quadratic_schedule(28, 0.025)
scheduler.set_timesteps(sigmas=sigmas)
print(scheduler.timesteps)

flip_scheduler = FlowMatchEulerDiscreteScheduler(invert_sigmas=True)
sigmas = _flip_linear_quadratic_schedule(28, 0.025)
flip_scheduler.set_timesteps(sigmas=sigmas)
print(flip_scheduler.timesteps)
print(torch.abs(scheduler.timesteps - flip_scheduler.timesteps).max())
```

Reply:
They are right now, because of this change: https://github.com/huggingface/diffusers/pull/9769/files#diff-2f110a1896ae6fef680356cbb9b0dcbdabd2634bf6e82e099291d8b8f35728a6R211

The differences is torch.ones vs torch.zeros. If we don't have the `inverse_sigmas` flag, we would end up appending `torch.zeros` to the inverted sigmas from the custom linear_quadractic_schedule, which is incorrect

Reply:
Let's say we remove the changes related to `invert_sigmas` completely in the Flow match scheduler.

- Using `dont_flip_linear_quadratic_schedule`: We will pass `timestep` to the transformer - which is what we want. We will also append `torch.zeros` to the sigmas in `FlowMatchScheduler.set_timestep`. This is incorrect because torch.zeros corresponds to complete noisy latent for this specific model based on my explanation a few comments above. Denoising won't work.
- Using `flip_linear_quadratic_schedule`: We will have to pass `1000 - timestep` because the model is associating to noise/clean latents on inverted timestep value. We will append `torch.zeros` in `FlowMatchScheduler.set_timestep` - which is correct this time. But, the denosing prediction computation in Mochi requires `sigma - sigma_next`, however we will do `sigma_next - sigma`, which is wrong again.

Reply:
Oh I see. Missed the `torch.ones` part in the scheduler. Cool 👍🏽